nixos/tests/vpp.nix

```diff
@@ -0,0 +1,135 @@
+# on another, and a VPP-powered router inbetween connecting the two.
+
+import ./make-test-python.nix (
+  { pkgs, lib, ... }:
+  {
+    name = "vpp";
+
+    meta.maintainers = with lib.maintainers; [ romner-set ];
+
+    nodes = {
+      client =
+        { ... }:
+        {
+          virtualisation.vlans = [ 1 ];
+          networking.interfaces = lib.mkForce {
+            eth1.ipv6.addresses = [
+              {
+                address = "fd01::2";
+                prefixLength = 64;
+              }
+            ];
+          };
+          networking.defaultGateway6 = "fd01::1";
+          networking.useDHCP = false;
+        };
+      server =
+        { ... }:
+        {
+          virtualisation.vlans = [ 2 ];
+          networking.interfaces = lib.mkForce {
+            eth1.ipv6.addresses = [
+              {
+                address = "fd02::2";
+                prefixLength = 64;
+              }
+            ];
+          };
+          networking.defaultGateway6 = "fd02::1";
+          networking.useDHCP = false;
+
+          services.httpd.enable = true;
+          services.httpd.adminAddr = "foo@example.org";
+
+          networking.firewall.enable = false;
+        };
+      router =
+        { config, ... }:
+        {
+          virtualisation.vlans = [
+            1
+            2
+          ];
+
+          # Disable netdevs so VPP can take over them
+          networking.useDHCP = false;
+          networking.interfaces = lib.mkForce { };
+
+          # install igb_uio driver
+          boot.extraModulePackages = with config.boot.kernelPackages; [ dpdk-kmods ];
+
+          virtualisation.memorySize = 4096;
+          services.vpp.hugepages = {
+            autoSetup = true;
+            count = 1024;
+          };
+
+          services.vpp.instances.main = {
+            enable = true;
+            kernelModule = "igb_uio";
+            settings = {
+              unix.cli-listen = "/run/vpp/cli.sock"; # override a default value
+
+              plugins = {
+                # Selectively enable only necessary plugins
+                plugin.default.disable = true;
+                plugin."dpdk_plugin.so".enable = true;
+              };
+
+              dpdk.dev."0000:00:09.0".name = "vlan1";
+              dpdk.dev."0000:00:0a.0".name = "vlan2";
+            };
+            startupConfig = ''
+              set int state vlan1 up
+              set int state vlan2 up
+              set int ip addr vlan1 fd01::1/64
+              set int ip addr vlan2 fd02::1/64
+            '';
+          };
+
+          environment.systemPackages = with pkgs; [ vpp ];
+        };
+    };
+
+    testScript = ''
+      start_all()
+
+      # wait for router
+      router.wait_for_unit("vpp-main.service")
+      router.wait_for_file("/run/vpp/cli.sock")
+
+      # make sure VPP initialized correctly
+      # the sleep is necessary since startupConfig executes *after* the service starts
+      router.sleep(2)
```

Comment:
This could easily fail on hydra. Is it not enough to wait for the socket above?

Reply:
VPP reads `startupConfig` right after actually starting up, the 4 commands should execute near-instantaneously so I don't think it'll be an issue with 2s. It does feel like a bit of a hack, but I can't come up with any better solution.